Stratis.Bitcoin.Features.IndexStore/IndexRepository.cs

```diff
@@ -152,44 +167,51 @@ private void DropIndex(string name, DBreeze.Transactions.Transaction transaction
         public Task<bool> CreateIndex(string name, bool multiValue, string builder, string[] dependencies = null)
         {
             Guard.NotNull(name, nameof(name));
+            Guard.NotNull(builder, nameof(builder));
 
             return this.session.Execute(() =>
             {
-                if (this.Indexes.ContainsKey(name))
-                    throw new IndexStoreException("The '" + name + "' index already exists");
+                this.CreateIndex(name, multiValue, builder, dependencies, this.session.Transaction);
 
-                var index = new Index(this, name, multiValue, builder, dependencies);
-                this.Indexes[name] = index;
+                this.session.Transaction.Commit();
 
-                var dbIndex = this.session.Transaction.Select<string, string>("Common", index.Table);
-                if (dbIndex.Exists)
-                    this.session.Transaction.RemoveAllKeys(index.Table, true);
+                return true;
+            });
+        }
 
-                if (!this.tableNames.Contains(index.Table))
-                    this.tableNames.Add(index.Table);
+        private void CreateIndex(string name, bool multiValue, string builder, string[] dependencies, DBreeze.Transactions.Transaction transaction)
+        {
+            if (this.Indexes.ContainsKey(name))
+                throw new IndexStoreException("The '" + name + "' index already exists");
 
-                var transactions = new List<(Transaction, Block)>();
-                foreach (var row in this.session.Transaction.SelectForward<byte[], Block>("Block"))
-                {
-                    var block = row.Value;
-                    var key = block.GetHash().ToBytes();
+            var index = new Index(this, name, multiValue, builder, dependencies);
+            this.Indexes[name] = index;
 
-                    foreach (var transaction in block.Transactions)
-                        transactions.Add((transaction, block));
+            var dbIndex = transaction.Select<string, string>("Common", index.Table);
+            if (dbIndex.Exists)
+                transaction.RemoveAllKeys(index.Table, true);
 
-                    if (transactions.Count >= 5000)
-                    {
-                        index.IndexTransactionDetails(transactions);
-                        transactions.Clear();
-                    }
-                }
+            if (!this.tableNames.Contains(index.Table))
+                this.tableNames.Add(index.Table);
 
-                index.IndexTransactionDetails(transactions);
-                this.session.Transaction.Insert<string, string>("Common", index.Table, index.ToString());
-                this.session.Transaction.Commit();
+            var transactions = new List<(Transaction, Block)>();
+            foreach (var row in transaction.SelectForward<byte[], Block>("Block"))
```

Comment:
Ah you have to reindex the entire store, this might be a very long operation no?

Reply:
It can be a very long operation on the first run, for a new index, unless the block table is empty at the time... On subsequent runs the index will already exist and will keep pace with changes. Nevertheless it may be possible to improve on this despite the single-threaded requirements. Such an approach would involve building the new index in the background and only committing it once it had been fully built.

Reply:
The above "improvement" must be weighed against the fact that we don't really want to start the "block explorer" node, including the block explorer, before all indexes are aligned with the tip... Also, unless I am mistaken, this situation can only arise during development anyway OR if the index folder is deleted without also deleting the blocks folder.

Reply:
That's fine its a long operation there is no escaping that we just need to make sure the node is displaying progress some how.